_posts/2023-04-02-writing-custom-pod-security-providers.md

```diff
@@ -0,0 +1,271 @@
+You can just extend the provider and re-implement only the methods you want to change.
+Strimzi currently includes two providers: [`BaselinePodSecurityProvider` and `RestrictedPodSecurityProvider`](https://github.com/strimzi/strimzi-kafka-operator/tree/main/api/src/main/java/io/strimzi/plugin/security/profiles/impl)
+
+For our example, we are going to take the [`RestrictedPodSecurityProvider`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/api/src/main/java/io/strimzi/plugin/security/profiles/impl/RestrictedPodSecurityProvider.java).
+It configures the security context of the Strimzi operands to match Kubernetes' _restricted_ security profile.
+When you try to use the Kafka Connect Build and its [Kaniko](https://github.com/GoogleContainerTools/kaniko) builder with this provider, it will [throw an exception](https://github.com/strimzi/strimzi-kafka-operator/blob/main/api/src/main/java/io/strimzi/plugin/security/profiles/impl/RestrictedPodSecurityProvider.java#L94-L103) because the Kaniko container builder does not work under the _restricted_ profile.
+Let's say you want to use the `RestrictedPodSecurityProvider` to secure the operand Pods, but you also want to use the Kafka Connect Build feature without any restrictions.
+
+In such a case, you can simply extend the `RestrictedPodSecurityProvider` and override the `kafkaConnectBuildContainerSecurityContext` method with your implementation, which will just let it run instead of throwing an exception:
+
+```java
+package cz.scholz.providers;
+
+import io.fabric8.kubernetes.api.model.SecurityContext;
+import io.strimzi.plugin.security.profiles.ContainerSecurityProviderContext;
+import io.strimzi.plugin.security.profiles.impl.RestrictedPodSecurityProvider;
+
+public class CustomPodSecurityProvider2 extends RestrictedPodSecurityProvider {
+    @Override
+    public SecurityContext kafkaConnectBuildContainerSecurityContext(ContainerSecurityProviderContext context) {
+        if (context != null
+                && context.userSuppliedSecurityContext() != null)    {
+            return context.userSuppliedSecurityContext();
+        } else {
+            return null;
+        }
+    }
+}
+```
+
+This way you achieved what you wanted and you did not need to implement all the methods from scratch.
+
+#### Service Loader configuration
+
+When Strimzi uses the Pod Security Providers, it is using the [Java ServiceLoad](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) to load the implementations.
+To allow Strimzi to load your custom provider, you have to create a provider configuration file.
+The provider configuration file must be part of the custom provider source code and be packaged into the JAR file.
+The file should be named `io.strimzi.plugin.security.profiles.PodSecurityProvider` (the name of the interface which it implements) and placed in the `resources/META-INF/services/` path.
```

Comment:
Newer versions of Java allow declaring this in a `module-info.java` module descriptor too, which you could mention, even if you use `META-INF` for the example.